<a href="https://colab.research.google.com/github/Logan-Stark/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [1]:
# YOUR WORK HERE
# loading in the neww dataframe
!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip'

--2020-06-09 22:10:36--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.3’

student.zip.3       100%[===================>]  20.00K  --.-KB/s    in 0.04s   

2020-06-09 22:10:36 (537 KB/s) - ‘student.zip.3’ saved [20478/20478]



In [2]:
# unzipping
!unzip 'student.zip'

Archive:  student.zip
replace student-mat.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace student-por.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace student-merge.R? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace student.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


In [0]:
# Importing libraries
from scipy import stats
import pandas as pd
import numpy as np


In [0]:
df = pd.read_csv('student-mat.csv', sep=';')

In [5]:
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [6]:
# YOUR WORK HERE
# looking at different kinds of columns

df['failures'].value_counts().sort_index()

0    312
1     50
2     17
3     16
Name: failures, dtype: int64

In [7]:
df['internet'].value_counts().sort_index()

no      66
yes    329
Name: internet, dtype: int64

In [8]:
df.describe()

age        Medu        Fedu  ...          G1          G2          G3
count  395.000000  395.000000  395.000000  ...  395.000000  395.000000  395.000000
mean    16.696203    2.749367    2.521519  ...   10.908861   10.713924   10.415190
std      1.276043    1.094735    1.088201  ...    3.319195    3.761505    4.581443
min     15.000000    0.000000    0.000000  ...    3.000000    0.000000    0.000000
25%     16.000000    2.000000    2.000000  ...    8.000000    9.000000    8.000000
50%     17.000000    3.000000    2.000000  ...   11.000000   11.000000   11.000000
75%     18.000000    4.000000    3.000000  ...   13.000000   13.000000   14.000000
max     22.000000    4.000000    4.000000  ...   19.000000   19.000000   20.000000

[8 rows x 16 columns]

In [9]:
# make a crosstab
crosstab = pd.crosstab(df['internet'], df['failures'])
crosstab.head()

failures    0   1   2   3
internet                 
no         47  13   2   4
yes       265  37  15  12

In [10]:
# Performing our test
chi2, p_value, dof, expected = stats.chi2_contingency(crosstab)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 4.796825454131462
p value 0.18729362004759095
degrees of freedom 3
expected frequencies table [[ 52.13164557   8.35443038   2.84050633   2.67341772]
 [259.86835443  41.64556962  14.15949367  13.32658228]]


# Concepts
**Null Hypothesis**: Internet and Failure are not associated and are Independent.

**Alternative Hypothesis**: Internet and Failure time are associated and are dependent.

**confidence level**: 95%

# results 
- From our Chi statistic of 4.79 and a p-value of .187 we ACCEPT the null hypothesis that Internet and Failure are Independent" and REJECT the Alternate hypothesis " Internet and Failure are dependent".

In [11]:
crosstab_3 = pd.crosstab(df['traveltime'], df['address'])
crosstab_3

address      R    U
traveltime         
1           35  222
2           34   73
3           14    9
4            5    3

it is evident from our results that 'studytime' and 'absecnes' are not associated

In [12]:
chi2, p_value, dof, expected = stats.chi2_contingency(crosstab_3)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 43.96115908406922
p value 1.5381274806593009e-09
degrees of freedom 3
expected frequencies table [[ 57.2556962  199.7443038 ]
 [ 23.83797468  83.16202532]
 [  5.12405063  17.87594937]
 [  1.78227848   6.21772152]]


# Concepts
**Null Hypothesis**: Address and Tavel time are not associated and are Independent.

**Alternative Hypothesis**: Address and Tavel time are associated and are dependent.

**confidence level**: 95%

# results 
- From our Chi statistic of 43.9 and a p-value close to 0 we REJECT the null hypothesis that " Address and travel time are Independent" and accept the Alternate hypothesis " address and traveltime are dependent".

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [0]:
# YOUR WORK HERE

In [14]:
# make crosstba of observed values and save them
observed = pd.crosstab( df['address'], df['traveltime'])
observed

traveltime    1   2   3  4
address                   
R            35  34  14  5
U           222  73   9  3

In [15]:
# add margins and save
observed_with_margins = pd.crosstab( df['address'], df['traveltime'] , margins=True)
observed_with_margins

traveltime    1    2   3  4  All
address                         
R            35   34  14  5   88
U           222   73   9  3  307
All         257  107  23  8  395

In [16]:
# Define our row totals 
row_totals = df['address'].value_counts().sort_index().values
row_totals

array([ 88, 307])

In [17]:
# define our column totals
col_totals = df['traveltime'].value_counts().sort_index().values
col_totals

array([257, 107,  23,   8])

In [18]:
# grabbing values from margins
col_totals = observed_with_margins.iloc[0:2, 4:5].values.T[0]
col_totals

array([ 88, 307])

In [19]:
# grabbing values from margins
row_totals = observed_with_margins.iloc[2:3, 0:4].values[0]
row_totals

array([257, 107,  23,   8])

In [20]:
# Defining sample size
sample_size = df.shape[0]

sample_size

395

In [21]:
# Using for loop to compile data
my_list = []

for i, row_total in enumerate(row_totals):
  row = []
  for j, column_total in enumerate(col_totals):
    expected_val = (( column_total*row_total)/ sample_size)
    row.append(expected_val)
  my_list.append(row)

my_list


[[57.255696202531645, 199.74430379746835],
 [23.837974683544303, 83.1620253164557],
 [5.124050632911392, 17.87594936708861],
 [1.7822784810126582, 6.217721518987342]]

In [22]:
expected

array([[ 57.2556962 , 199.7443038 ],
       [ 23.83797468,  83.16202532],
       [  5.12405063,  17.87594937],
       [  1.78227848,   6.21772152]])

In [23]:
expected  = np.array(my_list).T
expected

array([[ 57.2556962 ,  23.83797468,   5.12405063,   1.78227848],
       [199.7443038 ,  83.16202532,  17.87594937,   6.21772152]])

In [26]:
observed = np.array(observed)
observed

array([[ 35,  34,  14,   5],
       [222,  73,   9,   3]])

In [27]:
# perfoming chi test
chi2 = (((observed - expected)**2) / expected).sum()

chi2

43.96115908406922

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)